### ISSUE TITLE
libphidget22: Release into ROS2 rolling broken

### ISSUE THREAD

**Issue by mintar** [ros-drivers/phidgets_drivers#113]
**TL;DR:** @nuclearsandwich has to remove `core.autocrlf` from his `.gitconfig` and re-run `bloom-release`.

## Problem

The project fails to build on the build farm on ROS2 rolling, caused by a bad `bloom-release`, probably by @nuclearsandwich:

https://build.ros2.org/job/Rbin_uJ64__libphidget22__ubuntu_jammy_amd64__binary/3/consoleFull

Relevant part:

```
08:42:27 cd /tmp/binarydeb/ros-rolling-libphidget22-2.2.1/obj-x86_64-linux-gnu/libphidget22-src && patch -p1 < /tmp/binarydeb/ros-rolling-libphidget22-2.2.1/patch/libphidgets22-1.6.20210312-fix-warnings.patch
08:42:27 patching file configure
08:42:27 patching file src/class/dataadapter.c
08:42:27 patching file src/util/dataadaptersupport.c
08:42:27 Reversed (or previously applied) patch detected!  Assume -R? [n] 
08:42:27 Apply anyway? [n] 
08:42:27 Skipping patch.
08:42:27 1 out of 1 hunk ignored -- saving rejects to file src/util/dataadaptersupport.c.rej
08:42:27 make[4]: Leaving directory '/tmp/binarydeb/ros-rolling-libphidget22-2.2.1/obj-x86_64-linux-gnu'
08:42:27 make[4]: *** [CMakeFiles/EP_libphidget22.dir/build.make:124: EP_libphidget22-prefix/src/EP_libphidget22-stamp/EP_libphidget22-patch] Error 1
08:42:27 make[3]: *** [CMakeFiles/Makefile2:140: CMakeFiles/EP_libphidget22.dir/all] Error 2
08:42:27 make[3]: Leaving directory '/tmp/binarydeb/ros-rolling-libphidget22-2.2.1/obj-x86_64-linux-gnu'
08:42:27 make[2]: Leaving directory '/tmp/binarydeb/ros-rolling-libphidget22-2.2.1/obj-x86_64-linux-gnu'
08:42:27 make[2]: *** [Makefile:149: all] Error 2
08:42:27 dh_auto_build: error: cd obj-x86_64-linux-gnu && make -j1 VERBOSE=1 returned exit code 2
08:42:28 make[1]: Leaving directory '/tmp/binarydeb/ros-rolling-libphidget22-2.2.1'
08:42:28 make[1]: *** [debian/rules:38: override_dh_auto_build] Error 25
08:42:28 make: *** [debian/rules:22: build] Error 2
08:42:28 dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
```

I had this problem before, so I know what caused it. The patch file in the release repository is different from the one in the source repo.

The source file has some `^M` line endings, which are required:

https://github.com/ros-drivers/phidgets_drivers/blob/2.2.1/libphidget22/patch/libphidgets22-1.6.20210312-fix-warnings.patch

```
$ cat -v patch/libphidgets22-1.6.20210312-fix-warnings.patch                                                                                  
diff -urp libphidget22-1.6.20210312.orig/configure libphidget22-1.6.20210312/configure
--- libphidget22-1.6.20210312.orig/configure	2021-03-12 17:23:02.000000000 +0000
+++ libphidget22-1.6.20210312/configure	2021-03-22 13:55:52.553323821 +0000
@@ -5681,7 +5681,7 @@ esac
 fi
 
 : ${AR=ar}
-: ${AR_FLAGS=cru}
+: ${AR_FLAGS=cr}
 
 
 
diff -urp libphidget22-1.6.20210312.orig/src/class/dataadapter.c libphidget22-1.6.20210312/src/class/dataadapter.c
--- libphidget22-1.6.20210312.orig/src/class/dataadapter.c	2021-03-12 17:22:46.000000000 +0000
+++ libphidget22-1.6.20210312/src/class/dataadapter.c	2021-03-22 13:52:16.752083410 +0000
@@ -90,7 +90,7 @@ PhidgetDataAdapter_bridgeInput(PhidgetCh
 	uint32_t dataLen;
 	Phidget_DeviceID deviceID;
 	Phidget_DeviceClass deviceClass;
-	PhidgetDataAdapter_PacketErrorCode err;
+	uint32_t err;
 
 	ch = (PhidgetDataAdapterHandle)phid;
 
diff -urp libphidget22-1.6.20210312.orig/src/util/dataadaptersupport.c libphidget22-1.6.20210312/src/util/dataadaptersupport.c
--- libphidget22-1.6.20210312.orig/src/util/dataadaptersupport.c	2021-03-12 17:22:45.000000000 +0000
+++ libphidget22-1.6.20210312/src/util/dataadaptersupport.c	2021-03-22 13:53:10.348894764 +0000
@@ -859,49 +859,6 @@ PhidgetReturnCode sendI2CData(PhidgetCha
 	return sendDataBuffer(ch, totalCount, (const uint8_t *)buffer, bp, waitResposne);
 }^M
 ^M
-static PhidgetReturnCode^M
-waitForNAKClear(mosiop_t iop, PhidgetChannelHandle ch, uint32_t milliseconds) {^M
-	PhidgetDataAdapterSupportHandle dataAdapterSupport;^M
-	mostime_t duration;^M
-	mostime_t start;^M
-^M
-	assert(ch);^M
-	TESTATTACHED(ch);^M
-^M
-	dataAdapterSupport = DATAADAPTER_SUPPORT(ch);^M
-	assert(dataAdapterSupport);^M
-^M
-	start = 0; // make compiler happy^M
-^M
-	if (milliseconds)^M
-		start = mos_gettime_usec();^M
-^M
-	PhidgetLock(ch);^M
-	for (;;) {^M
-		if (dataAdapterSupport->nakFlag == 0) {^M
-			PhidgetUnlock(ch);^M
-			return (EPHIDGET_OK);^M
-		}^M
-^M
-		if (!_ISOPEN(ch)) {^M
-			PhidgetUnlock(ch);^M
-			return (MOS_ERROR(iop, EPHIDGET_CLOSED, "Channel was closed while waiting for clear to send."));^M
-		}^M
-^M
-		if (milliseconds) {^M
-			duration = (mos_gettime_usec() - start) / 1000;^M
-			if (duration >= milliseconds) {^M
-^M
-				PhidgetUnlock(ch);^M
-				return (EPHIDGET_TIMEOUT);^M
-			}^M
-			PhidgetTimedWait(ch, milliseconds - (uint32_t)duration);^M
-		} else {^M
-			PhidgetWait(ch);^M
-		}^M
-	}^M
-}
-
 static PhidgetReturnCode sendTXDataVINT(mosiop_t iop, PhidgetChannelHandle ch, uint8_t *buf, size_t packetLen, PhidgetTransaction *trans){
 	PhidgetReturnCode ret;
 	SetNAK(ch);
```

The release repo file is missing the line endings:

https://github.com/ros2-gbp/phidgets_drivers-release/blob/debian/ros-rolling-libphidget22_2.2.1-2_jammy/patch/libphidgets22-1.6.20210312-fix-warnings.patch

```
$ cat -v /tmp/phidgets_drivers-release/patch/libphidgets22-1.6.20210312-fix-warnings.patch              
diff -urp libphidget22-1.6.20210312.orig/configure libphidget22-1.6.20210312/configure
--- libphidget22-1.6.20210312.orig/configure	2021-03-12 17:23:02.000000000 +0000
+++ libphidget22-1.6.20210312/configure	2021-03-22 13:55:52.553323821 +0000
@@ -5681,7 +5681,7 @@ esac
 fi
 
 : ${AR=ar}
-: ${AR_FLAGS=cru}
+: ${AR_FLAGS=cr}
 
 
 
diff -urp libphidget22-1.6.20210312.orig/src/class/dataadapter.c libphidget22-1.6.20210312/src/class/dataadapter.c
--- libphidget22-1.6.20210312.orig/src/class/dataadapter.c	2021-03-12 17:22:46.000000000 +0000
+++ libphidget22-1.6.20210312/src/class/dataadapter.c	2021-03-22 13:52:16.752083410 +0000
@@ -90,7 +90,7 @@ PhidgetDataAdapter_bridgeInput(PhidgetCh
 	uint32_t dataLen;
 	Phidget_DeviceID deviceID;
 	Phidget_DeviceClass deviceClass;
-	PhidgetDataAdapter_PacketErrorCode err;
+	uint32_t err;
 
 	ch = (PhidgetDataAdapterHandle)phid;
 
diff -urp libphidget22-1.6.20210312.orig/src/util/dataadaptersupport.c libphidget22-1.6.20210312/src/util/dataadaptersupport.c
--- libphidget22-1.6.20210312.orig/src/util/dataadaptersupport.c	2021-03-12 17:22:45.000000000 +0000
+++ libphidget22-1.6.20210312/src/util/dataadaptersupport.c	2021-03-22 13:53:10.348894764 +0000
@@ -859,49 +859,6 @@ PhidgetReturnCode sendI2CData(PhidgetCha
 	return sendDataBuffer(ch, totalCount, (const uint8_t *)buffer, bp, waitResposne);
 }
 
-static PhidgetReturnCode
-waitForNAKClear(mosiop_t iop, PhidgetChannelHandle ch, uint32_t milliseconds) {
-	PhidgetDataAdapterSupportHandle dataAdapterSupport;
-	mostime_t duration;
-	mostime_t start;
-
-	assert(ch);
-	TESTATTACHED(ch);
-
-	dataAdapterSupport = DATAADAPTER_SUPPORT(ch);
-	assert(dataAdapterSupport);
-
-	start = 0; // make compiler happy
-
-	if (milliseconds)
-		start = mos_gettime_usec();
-
-	PhidgetLock(ch);
-	for (;;) {
-		if (dataAdapterSupport->nakFlag == 0) {
-			PhidgetUnlock(ch);
-			return (EPHIDGET_OK);
-		}
-
-		if (!_ISOPEN(ch)) {
-			PhidgetUnlock(ch);
-			return (MOS_ERROR(iop, EPHIDGET_CLOSED, "Channel was closed while waiting for clear to send."));
-		}
-
-		if (milliseconds) {
-			duration = (mos_gettime_usec() - start) / 1000;
-			if (duration >= milliseconds) {
-
-				PhidgetUnlock(ch);
-				return (EPHIDGET_TIMEOUT);
-			}
-			PhidgetTimedWait(ch, milliseconds - (uint32_t)duration);
-		} else {
-			PhidgetWait(ch);
-		}
-	}
-}
-
 static PhidgetReturnCode sendTXDataVINT(mosiop_t iop, PhidgetChannelHandle ch, uint8_t *buf, size_t packetLen, PhidgetTransaction *trans){
 	PhidgetReturnCode ret;
 	SetNAK(ch);
```

So how did these line endings disappear? Probably because whoever did the release (according to the git history I guess it's @nuclearsandwich) had the following in their `.gitconfig`:

```
[core]
	autocrlf = input
```

See https://github.com/ros-drivers/phidgets_drivers/issues/77#issuecomment-638766192 .

## Solution

@nuclearsandwich has to remove `core.autocrlf` from his `.gitconfig` and re-run `bloom-release`. It is probably not safe to have this setting enabled on any PC that runs `bloom-release`.

**Comment by nuclearsandwich** [comment]
Thanks for opening the issue. This is one of those situations where bloom's use of git porcelain causes some grief. To the best of my knowledge there isn't an easy way for me to use a fixed global gitconfig file other than my own ~/.gitconfig

Would you like me to push an updated release or wait for #115 and a maintainer release?

**Comment by mintar** [comment]
> This is one of those situations where bloom's use of git porcelain causes some grief.

Tell me about it... The first time this happened to me, it took me ages to find the bug. I couldn't believe that bloom had changed the code that I had committed. Line endings are especially nefarious because my favourite diff tool (`meld`) ignores line endings, so it didn't show any changes either...

> Would you like me to push an updated release or wait for https://github.com/ros-drivers/phidgets_drivers/pull/115 and a maintainer release?

No, we'll do a maintainer release. I've only come up with #115 after opening this issue.

I'll close this. Thanks for your quick response!
